File: eunji/0x08/BOJ_9012.cpp
#include <iostream>
#include <deque>
using namespace std;

string check(string paren){
    deque<char> D;
    for (auto c : paren){
        if (D.empty()){
            D.push_back(c);
        }
        else if (c == ')'){
            if (D.back() == '('){
                D.pop_back();
            }
            else D.push_back(c);
        }
        else D.push_back(c);
    }
    if (D.size() != 0) return "NO\n";
    else return "YES\n";
}


int main(void){
    cin.tie(0);
    cin.sync_with_stdio(0);
    int N;
    cin >> N;
    for (int i=0; i<N; i++){
        string paren;
        cin >> paren;
        cout << check(paren);
    }
}

File: eunji/0x08_stackApp/BOJ_9012.py
from collections import deque
import sys

def check(paren):
    D = deque([])
    for c in paren:
        if len(D) == 0: D.append(c)
        elif c == ')':
            if D[-1] == '(': D.pop()
            else: D.append(c)
        else: D.append(c)
    if len(D) != 0: return "NO"
    else: return "YES"


N = int(sys.stdin.readline())
for i in range(N):
    paren = sys.stdin.readline()[:-1]
    print(check(paren))
